### PR TITLE
Default OAuth2 Scope Validator implementation 

### PR DESCRIPTION
**Pull request by shashimalcse**
### Proposed changes in this pull request

Scope validator implementation for Identity Server new authorization runtime. 

### Related Issues :
 - https://github.com/wso2/product-is/issues/16363
-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java [hunk 1]
```diff
@@ -1188,4 +1221,63 @@ protected void unsetRealmService(RealmService realmService) {
...
+    @Reference(
+            name = "identity.authorized.api.management.component",
+            service = AuthorizedAPIManagementService.class,
+            cardinality = ReferenceCardinality.MANDATORY,
+            policy = ReferencePolicy.DYNAMIC,
+            unbind = "unsetAuthorizedAPIManagementService"
+    )
+    protected void setAuthorizedAPIManagementService(AuthorizedAPIManagementService authorizedAPIManagementService) {
+
+        OAuth2ServiceComponentHolder.getInstance()
```
Comment (shashimalcse) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponent.java:1234]: add debug logs

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java [hunk 2]
```diff
@@ -102,6 +105,10 @@ public class OAuth2ServiceComponentHolder {
     private RefreshTokenGrantProcessor refreshTokenGrantProcessor;
     private OAuth2RevocationProcessor revocationProcessor;
```
Comment (shashimalcse) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java:107]: put a comment here

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/validationhandler/ScopeValidationContext.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
+package org.wso2.carbon.identity.oauth2.validators.validationhandler;
```
Comment (shashimalcse) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/validationhandler/ScopeValidationContext.java:1]: add license headers

**Review by sadilchamishka — DISMISSED** [review]

**Comment by shashimalcse** [comment]
PR builder have failed due to the : 
```
 [ERROR] Failures: 
 [ERROR] org.wso2.identity.integration.test.scim2.SCIM2PaginationTestCase.testUserListingPagination(org.wso2.identity.integration.test.scim2.SCIM2PaginationTestCase)
 [ERROR] Run 1: SCIM2PaginationTestCase.testUserListingPagination:125->getCurrentUsersCount:282 » AxisFault
 [ERROR] Run 2: SCIM2PaginationTestCase.testUserListingPagination:125->getCurrentUsersCount:282 » AxisFault
 [INFO] 
 [ERROR] org.wso2.identity.integration.test.scim2.SCIM2PaginationTestCase.testUserListingPaginationWithoutParams(org.wso2.identity.integration.test.scim2.SCIM2PaginationTestCase)
 [ERROR] Run 1: SCIM2PaginationTestCase.testUserListingPaginationWithoutParams:157->getCurrentUsersCount:282 » AxisFault
 [ERROR] Run 2: SCIM2PaginationTestCase.testUserListingPaginationWithoutParams:157->getCurrentUsersCount:282 » AxisFault
 [INFO] 
 [ERROR] JDBCBasedUserMgtTestCase.testInit:39 » UserAdminUserAdmin UserAdminUserAdminEx...
```
Tested locally and build succeeded

**Review by AnuradhaSK — APPROVED** [review]

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java [hunk 4]
```diff
@@ -738,4 +744,52 @@ public void setAccessTokenProvider(AccessTokenProvider accessTokenProvider) {
...
+    }
+
+    /**
+     * Get APIResourceManager osgi service.
+     *
+     * @return APIResourceManager.
+     */
+    public APIResourceManager getApiResourceManager() {
+        return apiResourceManager;
+    }
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java:766]:
```suggestion
    public APIResourceManager getApiResourceManager() {
    
        return apiResourceManager;
    }
```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java [hunk 5]
```diff
@@ -738,4 +744,52 @@ public void setAccessTokenProvider(AccessTokenProvider accessTokenProvider) {
...
+    /**
+     * Set APIResourceManager osgi service.
+     *
+     * @param apiResourceManager APIResourceManager.
+     */
+    public void setApiResourceManager(APIResourceManager apiResourceManager) {
+        this.apiResourceManager = apiResourceManager;
+    }
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/internal/OAuth2ServiceComponentHolder.java:774]:
```suggestion
    public void setApiResourceManager(APIResourceManager apiResourceManager) {
    
        this.apiResourceManager = apiResourceManager;
    }
```

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/validationhandler/impl/RoleBasedScopeValidationHandler.java [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Role based scope validation handler validate scopes based on users roles.
+ */
+public class RoleBasedScopeValidationHandler implements ScopeValidationHandler {
+
+    private static final Log LOG = LogFactory.getLog(DefaultOAuth2ScopeValidator.class);
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/validationhandler/impl/RoleBasedScopeValidationHandler.java:43]: private static final Log LOG = LogFactory.getLog(RoleBasedScopeValidationHandler.class);

**Pull request merged**